src/components/dashboard/data/utils.js

```diff
@@ -1,4 +1,33 @@
-import { ASSIGNMENT_TYPES } from '../../enterprise-user-subsidy/enterprise-offers/data/constants';
+import { ASSIGNMENT_TYPES, ASSIGNMENT_ACTION_TYPES } from '../../enterprise-user-subsidy/enterprise-offers/data/constants';
+import {
+  LEARNER_ACKNOWLEDGED_ASSIGNMENT_CANCELLATION_ALERT,
+  LEARNER_ACKNOWLEDGED_ASSIGNMENT_EXPIRATION_ALERT,
+} from '../main-content/course-enrollments/data/constants';
+
+export function getIsActiveExpiredAssignment(assignments) {
+  const lastExpiredAlertDismissedTime = global.localStorage.getItem(LEARNER_ACKNOWLEDGED_ASSIGNMENT_EXPIRATION_ALERT);
+
+  const activeExpiredAssignments = assignments.filter((assignment) => (
+    assignment?.actions.some((action) => (
+      action.actionType === ASSIGNMENT_ACTION_TYPES.AUTOMATIC_CANCELLATION_NOTIFICATION
```

Comment:
Note: the cron job that runs to automatically cancel assignments will result in the assignment(s) being marked as `state=cancelled`, which means those assignment cards would actually render as the canceled assignment card variant instead of the expired variant for assignments where the cron job already ran.

I believe at this point, the handing of expired assignments is for the edge case where a learner is viewing an assignment in the UI that hasn't yet been canceled by the cron job. That said, I think the logic here may need to be based on the determined enrollment deadline as depicted by `isAssignmentExpired`, i.e. are there expired assignments that haven't been dismissed with an assignment enrollment deadline that is more recent than the last dismissed expired assignments?

[idea] Refactor `isAssignmentExpired` slightly to return both an `isExpired` boolean but also the enrollment deadline for the assignment and then check the localStorage date against this returned enrollment deadline.

```jsx
export const isAssignmentExpired = (assignment) => {
  // Assignment is not in an allocated state, so it cannot be expired.
  if (!assignment) {
    return false;
  }

  const currentDate = dayjs();
  const allocationDate = dayjs(assignment.created);
  const enrollmentEndDate = dayjs(assignment.contentMetadata.enrollByDate);
  const subsidyExpirationDate = dayjs(assignment.subsidyExpirationDate);

  const isExpired = (
    currentDate.diff(allocationDate, 'day') > 90
    || currentDate.isAfter(enrollmentEndDate)
    || currentDate.isAfter(subsidyExpirationDate)
  );

  const earliestAssignmentExpiryDate = [
    dayjs(allocationDate).add(90, 'day'),
    enrollmentEndDate,
    subsidyExpirationDate,
  ].sort()[0]?.toDate();

  return {
    isExpired,
    enrollByDeadline: earliestAssignmentExpiryDate,
  };
};
```

```jsx
export function getHasActiveExpiredAssignment(assignments) {
  const lastExpiredAlertDismissedTime = global.localStorage.getItem(LEARNER_ACKNOWLEDGED_ASSIGNMENT_EXPIRATION_ALERT);

  const activeExpiredAssignments = assignments.filter((assignment) => {
    const {
      isExpired,
      enrollByDeadline,
    } = isAssignmentExpired(assignment);

    if (!isExpired) {
      return false;
    }

    return dayjs(enrollByDeadline).isAfter(new Date(lastExpiredAlertDismissedTime));
  });

  return activeExpiredAssignments.length > 0;
}
```

Reply:
This is a good idea for now, but really we should introduce an `expired` state on the backend - and then the frontend could determine if an assignment is expired by checking `assignment.state === 'expired'`.